### PR TITLE
Initial code vewrsion.

### LINKED ISSUE #59: Export filtered-out methods to report file (not just stdout)
### Feature Description

Add the ability to export the list of filtered out methods to a file when running the tool in `--verify` or `--dry-run` mode, enabling downstream tools, CI logs, or users to consume this information in a structured way. The export should support at least one plain text or machine-readable (e.g., CSV or JSON) format.

Current behavior prints filtered methods to stdout only, based on compiled bytecode. A new CLI flag such as `--report-file <path>` should allow saving the output to the specified file, with clear, structured output matching what is currently printed to the console.

### Problem / Opportunity

Currently, users have no way to programmatically export which methods are filtered out except by redirecting stdout manually. This limits automation, traceability, and review processes in CI and makes it hard to integrate with other tooling. Automated review of which real (not synthetic/generated) methods are filtered becomes cumbersome, as the output is both ephemeral and not structured for parsing.

### Acceptance Criteria

1. User can run the CLI tool with a new flag (e.g., `--report-file <path>`) to save a list of filtered (and rescued) methods to a file.
2. Report output is the same as or clearer than the current console output (plain text and/or structured format like JSON/CSV).
3. If not specified, behavior should remain unchanged (output to stdout only).
4. Works for both `--verify` and `--dry-run` modes, across all supported integrations (CLI, sbt, Maven).
5. Well-documented in README and help output.

### Proposed Solution

Update `CoverageRewriter` and `VerifyScanner` so that the report output can be optionally written to a file, not just printed to console. Add a CLI flag (`--report-file <path>`, or similar) and document the format.

Where possible, support a structured data option (e.g. `--report-format csv|json|txt`), or at least document how the plain text format is structured for future extension.

Implementation would involve plumbing the output stream, not just println, to the appropriate sink.

### Dependencies / Related

Related to `CoverageRewriter`, `VerifyScanner`, CLI, sbt and Maven plugins. No direct blocking issues known.

### Additional Context

Reference: https://github.com/MoranaApps/jacoco-method-filter (README usage and verify/dry-run mode, examples of CLI and Maven tasks).

If exporting source-level info is considered for a future roadmap, note that the current tool analyzes compiled classes, not source files.

### PR DESCRIPTION
**Pull request by miroslavpojer**
### New features

- **Report file export** — filtered-methods report can now be written to a file after rewrite or verify.
  Configure via `--report-file <path>` and `--report-format txt|json|csv` (CLI),
  `jmfReportFile` / `jmfReportFormat` (sbt), or `jmf.reportFile` / `jmf.reportFormat` (Maven).
  Default format is `txt`; `json` and `csv` are also supported.
  Parent directories are created automatically if they do not exist.

Release Notes:
- Added `--report-file <path>` and `--report-format txt|json|csv` CLI flags to export the filtered-methods report to a file after rewrite or verify
- Added `jmfReportFile` and `jmfReportFormat` sbt settings (wired into `jmfVerify` and rewrite tasks)
- Added `jmf.reportFile` and `jmf.reportFormat` Maven parameters on both `rewrite` and `verify` mojos
- Default format is `txt`; `json` and `csv` are also supported; unknown format values are rejected with a validation error

Closes #59

### CODE REVIEW

**Review by miroslavpojer — COMMENTED** [review]
Self review

File: sbt-plugin/README.md [hunk 1]
```diff
@@ -160,6 +232,8 @@ Only configure the settings below when you want to:
 | `jmfLocalRulesFile` | `File` | `jmf-rules.txt` | Fallback local rules file used only when both `jmfGlobalRules` and `jmfLocalRules` are `None` |
 | `jmfDryRun` | `Boolean` | `false` | Dry run mode - logs matches without modifying classes |
 | `jmfOutDir` | `File` | `target` | Base output directory; filtered classes are written under `jmfOutDir / "classes-filtered"` |
+| `jmfReportFile` | `Option[File]` | `None` | Write a filtered-methods report to this file. Works with `jmfVerify` and when `jmfDryRun = true`. If not set, output goes to console only. |
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply sbt-plugin/README.md]: Fixed.

File: rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/CoverageRewriter.scala [hunk 2]
```diff
@@ -57,6 +61,11 @@ object CoverageRewriter {
     }
 
     println(s"[info] Processed $files class file(s), marked $marked method(s). dry-run=${cfg.dryRun}")
+
+    cfg.reportFile.foreach { path =>
+      val result = VerifyScanner.scan(cfg.in, rules)
+      writeReportFile(path, result.formatReport(cfg.reportFormat))
+    }
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/CoverageRewriter.scala:70]: Fixed.

File: rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/VerifyScanner.scala [hunk 3]
```diff
@@ -67,9 +65,88 @@ final case class ScanResult(
...
+    val excluded = excludedMethods
+    val rescued  = rescuedMethods
+    val sb       = new StringBuilder
+    sb.append("outcome,class,method,descriptor,exclusionRuleIds,inclusionRuleIds\n")
+    excluded.foreach { m =>
+      sb.append(s"EXCLUDED,${cell(m.fqcn)},${cell(m.methodName)},${cell(m.descriptor)},${cell(m.exclusionIds.mkString("|"))},\n")
+    }
+    rescued.foreach { m =>
+      sb.append(s"RESCUED,${cell(m.fqcn)},${cell(m.methodName)},${cell(m.descriptor)},${cell(m.exclusionIds.mkString("|"))},${cell(m.inclusionIds.mkString("|"))}\n")
+    }
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/VerifyScanner.scala]: Fixed.

File: rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/VerifyScanner.scala [hunk 4]
```diff
@@ -67,9 +65,88 @@ final case class ScanResult(
...
+    lines.mkString("\n")
+  }
+
+  private def formatJson(): String = {
+    def esc(s: String): String = s.replace("\\", "\\\\").replace("\"", "\\\"")
+    def str(s: String): String = s""""${esc(s)}""""
+    def strArr(seq: Seq[String]): String = seq.map(str).mkString("[", ", ", "]")
+
+    val excluded = excludedMethods
+    val rescued  = rescuedMethods
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/VerifyScanner.scala]: Fixed.

File: rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/CoverageRewriterCli.scala [hunk 5]
```diff
@@ -45,6 +45,20 @@ private[jacocomethodfilter] object CoverageRewriterCli {
...
+        .action((v, c) => c.copy(reportFile = Some(Paths.get(v))))
+        .text("Write filtered-methods report to this file (useful with --verify or --dry-run)")
+
+      opt[String]("report-format")
+        .optional()
+        .action((v, c) => c.copy(reportFormat = v))
+        .validate(v =>
+          if (Set("txt", "json", "csv").contains(v)) success
+          else failure("--report-format must be one of: txt, json, csv")
+        )
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/CoverageRewriterCli.scala:59]: Fixed.

File: integration-tests/test-sbt-report-file.sh [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+pass "$TEST_NAME — csv report file content correct"
+
+# ── 4. No report file written when jmfReportFile is None (default) ─────────
+info "$TEST_NAME — verifying no report file written when jmfReportFile = None"
+DEFAULT_OUT="$WORK_DIR/should-not-exist.txt"
+sbt jmfVerify > /dev/null 2>&1 || true
+if [[ -f "$DEFAULT_OUT" ]]; then
+  fail "$TEST_NAME — report file should not be written when jmfReportFile = None"
+fi
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply integration-tests/test-sbt-report-file.sh]: Fixed.

File: sbt-plugin/README.md [hunk 7]
```diff
@@ -101,6 +101,78 @@ Example output:
...
+    #productArity()I  rule-id:scala-product-arity
+    #productElement(I)Ljava/lang/Object;  rule-id:scala-product-element
+    #productIterator()Lscala/collection/Iterator;  rule-id:scala-product-iterator
+    #productPrefix()Ljava/lang/String;  rule-id:scala-product-prefix
+
+RESCUED by include rules (1 methods):
+  example.Calculator
+    #toString()Ljava/lang/String;  excl:scala-tostring → incl:rescue-calc-tostring
+
+Summary: 3 classes scanned, 10 methods excluded, 1 methods rescued
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply sbt-plugin/README.md]: Fixed.

File: README.md [hunk 8]
```diff
@@ -141,11 +141,24 @@ See plugin READMEs for full defaults and customization options.
...
+ data is missing:
 
 - [`sbt-plugin/README.md`](./sbt-plugin/README.md)
 - [`maven-plugin/README.md`](./maven-plugin/README.md)
 
+### Report File Export
+
+Add `--report-file <path>` (and optionally `--report-format txt|json|csv`) to any `--verify` or `--dry-run` invocation
+ to save filtered-method results to a file. This is useful for CI traceability and downstream automation. The file
+  content mirrors console output; JSON and CSV formats are also available for machine-readable consumption.
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply README.md]: Fixed.

File: rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/CoverageRewriterCli.scala [hunk 9]
```diff
@@ -45,6 +45,20 @@ private[jacocomethodfilter] object CoverageRewriterCli {
         .action((_, c) => c.copy(verify = true))
         .text("Read-only scan: list all methods that would be excluded by rules")
 
+      opt[String]("report-file")
+        .optional()
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/CoverageRewriterCli.scala:49]: Fixed.

File: rewriter-core/src/test/scala/io/moranaapps/jacocomethodfilter/VerifyScannerSpec.scala [hunk 10]
```diff
@@ -364,6 +364,197 @@ class VerifyScannerSpec extends AnyFunSuite {
...
+
+  test("formatReport csv escapes fields containing commas") {
+    // Descriptors can contain commas in edge cases; verify cell quoting
+    val matches = Seq(
+      MatchedMethod("com.example.Foo", "bar", "(Ljava/util/Map;)V", Excluded, Seq("r,id"), Seq.empty, Opcodes.ACC_PUBLIC)
+    )
+    val result = ScanResult(1, 1, matches)
+    val csv = result.formatReport("csv")
+    // rule ID "r,id" contains a comma → must be quoted in CSV
+    assert(csv.contains("\"r,id\""))
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply rewriter-core/src/test/scala/io/moranaapps/jacocomethodfilter/VerifyScannerSpec.scala]: Fixed.

File: rewriter-core/src/test/scala/io/moranaapps/jacocomethodfilter/VerifyScannerSpec.scala [hunk 11]
```diff
@@ -364,6 +364,186 @@ class VerifyScannerSpec extends AnyFunSuite {
...
+    assert(lines.length == 1)
+    assert(lines(0) == "outcome,class,method,descriptor,exclusionRuleIds,inclusionRuleIds")
+  }
+
+  test("formatReport defaults to txt for unknown format") {
+    val result = ScanResult(1, 0, Seq.empty)
+    val output = result.formatReport("unknown")
+    assert(output.contains("Summary:"))
+    assert(!output.contains("{"))
+  }
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply rewriter-core/src/test/scala/io/moranaapps/jacocomethodfilter/VerifyScannerSpec.scala]: Fixed.

File: rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/VerifyScanner.scala [hunk 12]
```diff
@@ -67,8 +70,87 @@ final case class ScanResult(
...
+
+    out(s"[verify] Summary: ${plural(classesScanned, "class")} scanned, ${plural(excluded.size, "method")} excluded, ${plural(rescued.size, "method")} rescued")
+  }
+
+  /** Format the report as a string in the specified format: txt (default), json, or csv. */
+  def formatReport(format: String): String = format.toLowerCase match {
+    case "json" => formatJson()
+    case "csv"  => formatCsv()
+    case _      => formatTxt()
+  }
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/VerifyScanner.scala]: Fixed.

File: rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/CoverageRewriterCli.scala [hunk 13]
```diff
@@ -45,13 +45,29 @@ private[jacocomethodfilter] object CoverageRewriterCli {
...
+
       checkConfig { cfg =>
         if (!Files.isDirectory(cfg.in)) {
           failure("--in must exist and be a directory")
         } else if (!cfg.verify && cfg.out.isEmpty) {
           failure("--out is required when not in verify mode")
         } else if (cfg.globalRules.isEmpty && cfg.localRules.isEmpty) {
           failure("At least one of --global-rules or --local-rules must be specified")
+        } else if (cfg.reportFile.exists(Files.isDirectory(_))) {
+          failure("--report-file must be a file path, not an existing directory")
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply rewriter-core/src/main/scala/io/moranaapps/jacocomethodfilter/CoverageRewriterCli.scala:70]: Fixed.

File: README.md [hunk 14]
```diff
@@ -141,11 +141,24 @@ See plugin READMEs for full defaults and customization options.
...
 
 - [`sbt-plugin/README.md`](./sbt-plugin/README.md)
 - [`maven-plugin/README.md`](./maven-plugin/README.md)
 
+### Report File Export
+
+Add `--report-file <path>` (and optionally `--report-format txt|json|csv`) to any `--verify` or `--dry-run` invocation
+to save filtered-method results to a file. This is useful for CI traceability and downstream automation. The file
+content mirrors console output; JSON and CSV formats are also available for machine-readable consumption.
+
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply README.md]: Fixed.
